File: packages/ts/src/bip340Schnorr.ts
// SPDX-FileCopyrightText: Copyright © 2026 ReallyMe LLC. All rights reserved
//
// SPDX-License-Identifier: Apache-2.0

import { schnorr, secp256k1 } from "@noble/curves/secp256k1.js";
import { ReallyMeCryptoError } from "./errors.js";

export const BIP340_SCHNORR_SECRET_KEY_LENGTH = 32;
export const BIP340_SCHNORR_PUBLIC_KEY_LENGTH = 32;
export const BIP340_SCHNORR_MESSAGE_LENGTH = 32;
export const BIP340_SCHNORR_AUX_RAND_LENGTH = 32;
export const BIP340_SCHNORR_SIGNATURE_LENGTH = 64;

/**
 * BIP-340 Schnorr over secp256k1 backed by @noble/curves.
 *
 * BIP-340 signs a 32-byte message representative and requires 32 bytes of
 * auxiliary randomness. Keeping `auxRand32` explicit matches the Rust, Swift,
 * and Kotlin lanes and avoids a hidden provider-specific randomness policy at
 * the generic facade boundary.
 */
export const ReallyMeBip340Schnorr = {
  generateKeyPair(): { publicKey: Uint8Array; secretKey: Uint8Array } {
    const secretKey = secp256k1.utils.randomSecretKey();
    return {
      publicKey: schnorr.getPublicKey(secretKey),
      secretKey,
    };
  },

  deriveKeyPair(secretKey: Uint8Array): { publicKey: Uint8Array; secretKey: Uint8Array } {
    return {
      publicKey: this.derivePublicKey(secretKey),
      secretKey: secretKey.slice(),
    };
  },

  derivePublicKey(secretKey: Uint8Array): Uint8Array {
    if (secretKey.length !== BIP340_SCHNORR_SECRET_KEY_LENGTH) {
      throw new ReallyMeCryptoError("invalid-input");
    }
    try {
      return schnorr.getPublicKey(secretKey);
    } catch {
      throw new ReallyMeCryptoError("invalid-input");
    }
  },

  sign(
    message32: Uint8Array,
    secretKey: Uint8Array,
    auxRand32: Uint8Array,
  ): Uint8Array {
    if (
      message32.length !== BIP340_SCHNORR_MESSAGE_LENGTH ||
      secretKey.length !== BIP340_SCHNORR_SECRET_KEY_LENGTH ||
      auxRand32.length !== BIP340_SCHNORR_AUX_RAND_LENGTH
    ) {
      throw new ReallyMeCryptoError("invalid-input");
    }
    try {
      return schnorr.sign(message32, secretKey, auxRand32);
    } catch {
      throw new ReallyMeCryptoError("invalid-input");
    }
  },

  verify(
    signature: Uint8Array,
    message32: Uint8Array,
    publicKey: Uint8Array,
  ): void {
    if (
      signature.length !== BIP340_SCHNORR_SIGNATURE_LENGTH ||
      message32.length !== BIP340_SCHNORR_MESSAGE_LENGTH ||
      publicKey.length !== BIP340_SCHNORR_PUBLIC_KEY_LENGTH
    ) {
      throw new ReallyMeCryptoError("invalid-input");
    }
    try {
      if (!schnorr.verify(signature, message32, publicKey)) {
        throw new ReallyMeCryptoError("invalid-signature");
      }
    } catch {
      throw new ReallyMeCryptoError("invalid-signature");
    }
  },
} as const;

File: packages/kotlin/src/main/kotlin/me/really/crypto/Bip340Schnorr.kt
// SPDX-FileCopyrightText: Copyright © 2026 ReallyMe LLC. All rights reserved
//
// SPDX-License-Identifier: Apache-2.0

package me.really.crypto

import fr.acinq.secp256k1.Secp256k1
import fr.acinq.secp256k1.Secp256k1Exception

/**
 * BIP-340 Schnorr signatures over secp256k1, backed by Bitcoin Core
 * libsecp256k1 (the constant-time reference implementation) via ACINQ's JNI
 * bindings — the same C library the Swift lane uses through CSecp256k1. No EC
 * scalar math is hand-rolled on the JVM, so there is no secret-dependent
 * timing surface, and signatures are byte-identical to every other lane.
 *
 * The generic `ReallyMeCrypto.sign` facade does not route here because BIP-340
 * signing requires caller-controlled 32-byte auxiliary randomness. Callers use
 * this object directly when they need signing, and the generic facade exposes
 * key generation plus verification.
 */
public object ReallyMeBip340Schnorr {
    public const val SECRET_KEY_LENGTH: Int = 32
    public const val PUBLIC_KEY_LENGTH: Int = 32
    public const val MESSAGE_LENGTH: Int = 32
    public const val AUX_RAND_LENGTH: Int = 32
    public const val SIGNATURE_LENGTH: Int = 64

    public fun generateKeyPair(): Pair<ByteArray, ByteArray> {
        val secp = provider()
        return withRandomSecretCandidate(
            length = SECRET_KEY_LENGTH,
            isValid = secp::secKeyVerify,
        ) { secretKey ->
            Pair(derivePublicKey(secretKey), secretKey.copyOf())
        }
    }

    public fun deriveKeyPair(secretKey: ByteArray): Pair<ByteArray, ByteArray> =
        Pair(derivePublicKey(secretKey), secretKey.copyOf())

    /** Derives the 32-byte x-only BIP-340 public key for a 32-byte secret. */
    public fun derivePublicKey(secretKey: ByteArray): ByteArray {
        val secp = provider()
        validateSecretKey(secp, secretKey)
        return try {
            xOnly(secp, secp.pubkeyCreate(secretKey))
        } catch (_: Secp256k1Exception) {
            throw ReallyMeCryptoException.InvalidInput()
        }
    }

    public fun sign(message32: ByteArray, secretKey: ByteArray, auxRand32: ByteArray): ByteArray {
        if (message32.size != MESSAGE_LENGTH || auxRand32.size != AUX_RAND_LENGTH) {
            throw ReallyMeCryptoException.InvalidInput()
        }
        val secp = provider()
        validateSecretKey(secp, secretKey)
        return try {
            secp.signSchnorr(message32, secretKey, auxRand32)
        } catch (_: Secp256k1Exception) {
            throw ReallyMeCryptoException.ProviderFailure()
        }
    }

    public fun verify(signature: ByteArray, message32: ByteArray, publicKey: ByteArray) {
        if (
            signature.size != SIGNATURE_LENGTH ||
            message32.size != MESSAGE_LENGTH ||
            publicKey.size != PUBLIC_KEY_LENGTH
        ) {
            throw ReallyMeCryptoException.InvalidInput()
        }
        val secp = provider()
        val valid = try {
            secp.verifySchnorr(signature, message32, publicKey)
        } catch (_: Secp256k1Exception) {
            throw ReallyMeCryptoException.InvalidSignature()
        }
        if (!valid) {
            throw ReallyMeCryptoException.InvalidSignature()
        }
    }

    private fun provider(): Secp256k1 =
        try {
            Secp256k1.get()
        } catch (_: LinkageError) {
            throw ReallyMeCryptoException.ProviderFailure()
        } catch (_: RuntimeException) {
            throw ReallyMeCryptoException.ProviderFailure()
        }

    private fun validateSecretKey(secp: Secp256k1, secretKey: ByteArray) {
        if (secretKey.size != SECRET_KEY_LENGTH || !secp.secKeyVerify(secretKey)) {
            throw ReallyMeCryptoException.InvalidInput()
        }
    }

    /** x-only public key = the X coordinate of the SEC1 compressed public key. */
    private fun xOnly(secp: Secp256k1, uncompressedPublicKey: ByteArray): ByteArray =
        secp.pubKeyCompress(uncompressedPublicKey).copyOfRange(1, 1 + PUBLIC_KEY_LENGTH)
}
